src/provider/shplonk.rs

```diff
@@ -284,23 +263,41 @@ where
     // P_i+1(r^2) == (1 - point_i) * P_i_even + point_i * P_i_odd -> should hold, according to Gemini transformation
     let mut point = point.to_vec();
     point.reverse();
+
+    let r_mul_2 = E::Fr::from(2) * r;
     #[allow(clippy::disallowed_methods)]
-    for (index, ((eval_r, eval_minus_r), eval_r_squared)) in evals_at_r
-      .iter()
-      .zip_eq(evals_at_minus_r.iter())
-      // TODO: Ask Adrian if we need evals_at_r_squared[0] for some additional checks
-      .zip(evals_at_r_squared[1..].iter())
+    let verification_failed = pi.evals[0]
+      .par_iter()
+      .chain(&[*P_of_x])
+      .zip_eq(pi.evals[1].par_iter().chain(&[*P_of_x]))
+      .zip(pi.evals[2][1..].par_iter().chain(&[*P_of_x]))
       .enumerate()
-    {
-      let even = (*eval_r + eval_minus_r) * (E::Fr::from(2).invert().unwrap());
-      let odd = (*eval_r - eval_minus_r) * ((E::Fr::from(2) * r).invert().unwrap());
+      .any(|(index, ((eval_r, eval_minus_r), eval_r_squared))| {
+        // some optimisation to avoid using expensive inversions:
+        // P_i+1(r^2) == (1 - point_i) * (P_i(r) + P_i(-r)) * 1/2 + point_i * (P_i(r) - P_i(-r)) * 1/2 * r
+        // is equivalent to:
+        // 2 * r * P_i+1(r^2) == r * (1 - point_i) * (P_i(r) + P_i(-r)) + point_i * (P_i(r) - P_i(-r))
+
+        let even = *eval_r + eval_minus_r;
+        let odd = *eval_r - eval_minus_r;
+        let right = r * ((E::Fr::ONE - point[index]) * even) + (point[index] * odd);
+        let left = *eval_r_squared * r_mul_2;
+        left != right
+      });
 
-      if *eval_r_squared != ((E::Fr::ONE - point[index]) * even) + (point[index] * odd) {
-        return Err(NovaError::ProofVerifyError);
-      }
+    if verification_failed {
+      return Err(NovaError::ProofVerifyError);
     }
 
-    let C_P: E::G1 = pi.comms.iter().map(|comm| comm.to_curve()).rlc(&q);
+    // TODO: revisit C_P computing later with progress in https://github.com/huitseeker/arecibo/tree/rayon-parscan
+    let q_powers = HyperKZG::<E, NE>::batch_challenge_powers(q, comms.len());
```

Comment:
You can make `spartan::powers` `pub (in crate)`: 

```suggestion
    let q_powers = crate::spartan::powers(q, comms.len());
```